src/components/MetadataEditor/MetadataEditor.vue

```diff
@@ -0,0 +1,383 @@
+        <form-section v-bind="sectionBinds('metaspace_options')" v-on="sectionEvents('metaspace_options')"/>
+        <form-section v-for="sectionKey in otherSections"
+                      :key="sectionKey"
+                      v-bind="sectionBinds(sectionKey)"
+                      v-on="sectionEvents(sectionKey)"/>
+      </div>
+    </div>
+    <div id="load-indicator" v-else v-loading="true">
+    </div>
+  </div>
+</template>
+
+<script>
+ /*
+    This component serves two purposes:
+  * editing metadata of existing datasets;
+  * providing metadata during the initial upload.
+
+    It has a few simplifying assumptions on the structure and types used:
+  * nesting is limited to 2 levels: section -> field (any sub-fields are handled by having complex editors on the field level)
+  * sections are assumed to be objects (except Data_Type, which is a single string value)
+
+    If datasetId is provided, the component fetches existing metadata for
+    that dataset from the GraphQL server so that it can be edited.
+    Autocompletion functionality also relies on the GraphQL server.
+
+    The other two props are for controlling the submit button behavior.
+
+    On submit button click, the form is checked for validity; if valid,
+    a submit event is emitted with dataset ID and stringified form value.
+  */
+
+ import {defaultMetadataType, metadataSchemas} from '../../assets/metadataRegistry';
+ import {deriveFullSchema} from './formStructure';
+ import {get, set, isArray, isEqual, isPlainObject, mapValues, forEach, without} from 'lodash-es';
+ import {
+   fetchAutocompleteSuggestionsQuery,
+   fetchMetadataQuery,
+   metadataOptionsQuery
+ } from '../../api/metadata';
+ import DatabaseDescriptions from '../DatabaseDescriptions.vue';
+ import FormSection from './FormSection.vue';
+
+ const factories = {
+   'string': schema => schema.default || '',
+   'number': schema => schema.default || 0,
+   'object': schema => mapValues(schema.properties, prop => factories[prop.type](prop)),
+   'array': schema => schema.default || [],
+   'boolean': schema => schema.default || false,
+ };
+
+ const LOCAL_STORAGE_KEY = 'latestMetadataSubmission';
+ const LOCAL_STORAGE_VERSION_KEY = 'latestMetadataSubmissionVersion';
+
+ // TODO: fill in institution automatically when user profiles are added
+
+ export default {
+   name: 'metadata-editor',
+   props: {
+     datasetId: String,
+     enableSubmit: Boolean,
+     disabledSubmitMessage: String,
+     validationErrors: Array,
+   },
+   components: {
+     FormSection,
+   },
+
+   created() {
+     this.loadForm();
+   },
+
+   data() {
+     return {
+       value: null,
+       isPublic: true,
+       schema: null,
+       molecularDatabases: [],
+       possibleAdducts: {},
+     }
+   },
+
+   watch: {
+     '$store.getters.filter.metadataType'(newMdType) {
+       if (this.isNew && newMdType !== this.value.Data_Type) {
+         this.saveForm();
+         this.loadForm();
+       }
+     }
```

Comment:
This `watch` detects when you change `metadataType` on the upload form. I didn't spend much time on this workflow as we don't allow changing `metadataType` in prod yet, so if you put values in fields that don't exist in both LC-MS and IMS metadata, then those fields' values will be reset each time you change between the two types. The important piece for me here is that you can't accidentally submit fields from the LC-MS metadata as part of an IMS submission or vice versa.